File: oee-app/oee-interface/lib/domain/models/OEE_Calculation.ts
import { Availability } from './Availability'
import { Performance } from './Performance'
import { Quality } from './Quality'
import { oeeConfig } from '../../../config/oee.config'

/**
 * OEE_Calculation Value Object
 * 
 * Represents a complete OEE (Overall Equipment Effectiveness) calculation
 * following the canonical model structure.
 * 
 * OEE = Availability × Performance × Quality
 * 
 * This is the key metric for measuring manufacturing productivity, 
 * identifying losses, and improving equipment efficiency.
 */
export class OEE_Calculation {
  readonly oee_id: string
  readonly resource_reference: string
  readonly calculation_period_start: Date
  readonly calculation_period_end: Date
  
  private readonly availability: Availability
  private readonly performance: Performance
  private readonly quality: Quality

  constructor(params: {
    oee_id?: string
    resource_reference: string
    calculation_period_start: Date
    calculation_period_end: Date
    availability: Availability
    performance: Performance
    quality: Quality
  }) {
    if (params.calculation_period_end <= params.calculation_period_start) {
      throw new Error('Calculation period end must be after start')
    }

    this.oee_id = params.oee_id ?? `OEE-${params.resource_reference}-${Date.now()}`
    this.resource_reference = params.resource_reference
    this.calculation_period_start = params.calculation_period_start
    this.calculation_period_end = params.calculation_period_end
    this.availability = params.availability
    this.performance = params.performance
    this.quality = params.quality
  }

  /**
   * Get availability percentage (0-100)
   */
  get availability_percentage(): number {
    return this.availability.percentage
  }

  /**
   * Get performance percentage (0-100)
   */
  get performance_percentage(): number {
    return this.performance.percentage
  }

  /**
   * Get quality percentage (0-100)
   */
  get quality_percentage(): number {
    return this.quality.percentage
  }

  /**
   * Calculate overall OEE percentage (0-100)
   * OEE = Availability × Performance × Quality
   */
  get oee_percentage(): number {
    return (
      this.availability.decimal * 
      this.performance.decimal * 
      this.quality.decimal * 
      100
    )
  }

  /**
   * Get OEE as a decimal (0-1)
   */
  get oee_decimal(): number {
    return this.oee_percentage / 100
  }

  /**
   * Determine if OEE requires attention based on thresholds
   */
  requiresAttention(thresholds?: {
    oee?: number
    availability?: number
    performance?: number
    quality?: number
  }): boolean {
    const defaults = oeeConfig.alerts.requiresAttention
    const limits = { ...defaults, ...thresholds }

    return (
      this.oee_percentage < limits.oee ||
      this.availability_percentage < limits.availability ||
      this.performance_percentage < limits.performance ||
      this.quality_percentage < limits.quality
    )
  }

  /**
   * Identify which factor is the worst (constraining factor)
   */
  getWorstFactor(): 'availability' | 'performance' | 'quality' {
    const factors = {
      availability: this.availability_percentage,
      performance: this.performance_percentage,
      quality: this.quality_percentage
    }

    return Object.entries(factors).reduce((worst, [factor, value]) => 
      value < factors[worst as keyof typeof factors] ? factor : worst
    , 'availability') as 'availability' | 'performance' | 'quality'
  }

  /**
   * Get improvement potential for each factor
   */
  getImprovementPotential(worldClassTargets?: {
    availability?: number
    performance?: number
    quality?: number
  }): {
    availability: number
    performance: number
    quality: number
    overall: number
  } {
    const targets = {
      availability: worldClassTargets?.availability ?? oeeConfig.thresholds.availability.excellent,
      performance: worldClassTargets?.performance ?? oeeConfig.thresholds.performance.excellent,
      quality: worldClassTargets?.quality ?? oeeConfig.thresholds.quality.excellent
    }

    return {
      availability: Math.max(0, targets.availability - this.availability_percentage),
      performance: Math.max(0, targets.performance - this.performance_percentage),
      quality: Math.max(0, targets.quality - this.quality_percentage),
      overall: Math.max(0, (targets.availability * targets.performance * targets.quality / 10000) - this.oee_percentage)
    }
  }

  /**
   * Calculate what OEE would be if one factor improved
   */
  simulateImprovement(
    factor: 'availability' | 'performance' | 'quality',
    newPercentage: number
  ): number {
    const factors = {
      availability: factor === 'availability' ? newPercentage : this.availability_percentage,
      performance: factor === 'performance' ? newPercentage : this.performance_percentage,
      quality: factor === 'quality' ? newPercentage : this.quality_percentage
    }

    return (factors.availability * factors.performance * factors.quality) / 10000
  }

  /**
   * Get OEE classification based on industry standards
   */
  getClassification(): string {
    const oee = this.oee_percentage

    if (oee >= oeeConfig.thresholds.oee.excellent) {
      return 'World Class - Excellent performance'
    } else if (oee >= oeeConfig.thresholds.oee.good) {
      return 'Good - Acceptable performance'
    } else if (oee >= 40) {
      return 'Fair - Improvement needed'
    } else {
      return 'Poor - Significant improvement required'
    }
  }

  /**
   * Get detailed breakdown of all components
   */
  getBreakdown(): {
    oee: number
    availability: ReturnType<Availability['getBreakdown']>
    performance: ReturnType<Performance['getBreakdown']>
    quality: ReturnType<Quality['getBreakdown']>
    worstFactor: string
    classification: string
    periodHours: number
  } {
    const periodMs = this.calculation_period_end.getTime() - this.calculation_period_start.getTime()
    const periodHours = periodMs / (1000 * 60 * 60)

    return {
      oee: this.oee_percentage,
      availability: this.availability.getBreakdown(),
      performance: this.performance.getBreakdown(),
      quality: this.quality.getBreakdown(),
      worstFactor: this.getWorstFactor(),
      classification: this.getClassification(),
      periodHours
    }
  }

  /**
   * Create a summary object for reporting
   */
  toSummary(): {
    oee_id: string
    resource_reference: string
    period_start: string
    period_end: string
    oee_percentage: number
    availability_percentage: number
    performance_percentage: number
    quality_percentage: number
    worst_factor: string
    requires_attention: boolean
  } {
    return {
      oee_id: this.oee_id,
      resource_reference: this.resource_reference,
      period_start: this.calculation_period_start.toISOString(),
      period_end: this.calculation_period_end.toISOString(),
      oee_percentage: this.oee_percentage,
      availability_percentage: this.availability_percentage,
      performance_percentage: this.performance_percentage,
      quality_percentage: this.quality_percentage,
      worst_factor: this.getWorstFactor(),
      requires_attention: this.requiresAttention()
    }
  }

  /**
   * Value object equality
   */
  equals(other: OEE_Calculation): boolean {
    return (
      this.oee_id === other.oee_id &&
      this.resource_reference === other.resource_reference &&
      this.calculation_period_start.getTime() === other.calculation_period_start.getTime() &&
      this.calculation_period_end.getTime() === other.calculation_period_end.getTime() &&
      this.availability.equals(other.availability) &&
      this.performance.equals(other.performance) &&
      this.quality.equals(other.quality)
    )
  }

  toString(): string {
    return `OEE: ${this.oee_percentage.toFixed(1)}% (A:${this.availability_percentage.toFixed(1)}% × P:${this.performance_percentage.toFixed(1)}% × Q:${this.quality_percentage.toFixed(1)}%)`
  }
}

File: src/Industrial.Adam.Oee/Domain/Entities/OeeCalculation.cs
using Industrial.Adam.Oee.Domain.Interfaces;
using Industrial.Adam.Oee.Domain.ValueObjects;

namespace Industrial.Adam.Oee.Domain.Entities;

/// <summary>
/// OEE Calculation Aggregate Root
/// 
/// Represents a complete OEE (Overall Equipment Effectiveness) calculation
/// following the canonical model structure.
/// 
/// OEE = Availability × Performance × Quality
/// 
/// This is the key metric for measuring manufacturing productivity, 
/// identifying losses, and improving equipment efficiency.
/// </summary>
public sealed class OeeCalculation : Entity<string>, IAggregateRoot
{
    /// <summary>
    /// Resource reference (device/machine identifier)
    /// </summary>
    public string ResourceReference { get; private set; }

    /// <summary>
    /// Start of the calculation period
    /// </summary>
    public DateTime CalculationPeriodStart { get; private set; }

    /// <summary>
    /// End of the calculation period
    /// </summary>
    public DateTime CalculationPeriodEnd { get; private set; }

    /// <summary>
    /// Availability component of OEE
    /// </summary>
    public Availability Availability { get; private set; }

    /// <summary>
    /// Performance component of OEE
    /// </summary>
    public Performance Performance { get; private set; }

    /// <summary>
    /// Quality component of OEE
    /// </summary>
    public Quality Quality { get; private set; }

    /// <summary>
    /// When this calculation was created
    /// </summary>
    public DateTime CreatedAt { get; private set; }

    /// <summary>
    /// Default constructor for serialization
    /// </summary>
    private OeeCalculation() : base()
    {
        ResourceReference = string.Empty;
        Availability = new Availability(0, 0);
        Performance = new Performance(0, 0, 1);
        Quality = new Quality(0, 0);
    }

    /// <summary>
    /// Creates a new OEE calculation
    /// </summary>
    /// <param name="oeeId">Optional OEE calculation ID (generated if not provided)</param>
    /// <param name="resourceReference">Resource/device identifier</param>
    /// <param name="calculationPeriodStart">Start of calculation period</param>
    /// <param name="calculationPeriodEnd">End of calculation period</param>
    /// <param name="availability">Availability value object</param>
    /// <param name="performance">Performance value object</param>
    /// <param name="quality">Quality value object</param>
    /// <exception cref="ArgumentException">Thrown when parameters are invalid</exception>
    public OeeCalculation(
        string? oeeId,
        string resourceReference,
        DateTime calculationPeriodStart,
        DateTime calculationPeriodEnd,
        Availability availability,
        Performance performance,
        Quality quality) : base(oeeId ?? GenerateOeeId(resourceReference))
    {
        if (string.IsNullOrWhiteSpace(resourceReference))
            throw new ArgumentException("Resource reference cannot be null or empty", nameof(resourceReference));

        if (calculationPeriodEnd <= calculationPeriodStart)
            throw new ArgumentException("Calculation period end must be after start", nameof(calculationPeriodEnd));

        ResourceReference = resourceReference;
        CalculationPeriodStart = calculationPeriodStart;
        CalculationPeriodEnd = calculationPeriodEnd;
        Availability = availability ?? throw new ArgumentNullException(nameof(availability));
        Performance = performance ?? throw new ArgumentNullException(nameof(performance));
        Quality = quality ?? throw new ArgumentNullException(nameof(quality));
        CreatedAt = DateTime.UtcNow;
    }

    /// <summary>
    /// Get availability percentage (0-100)
    /// </summary>
    public decimal AvailabilityPercentage => Availability.Percentage;

    /// <summary>
    /// Get performance percentage (0-100)
    /// </summary>
    public decimal PerformancePercentage => Performance.Percentage;

    /// <summary>
    /// Get quality percentage (0-100)
    /// </summary>
    public decimal QualityPercentage => Quality.Percentage;

    /// <summary>
    /// Calculate overall OEE percentage (0-100)
    /// OEE = Availability × Performance × Quality
    /// </summary>
    public decimal OeePercentage =>
        Availability.Decimal * Performance.Decimal * Quality.Decimal * 100;

    /// <summary>
    /// Get OEE as a decimal (0-1)
    /// </summary>
    public decimal OeeDecimal => OeePercentage / 100;

    /// <summary>
    /// Calculate the period duration in hours
    /// </summary>
    public decimal PeriodHours
    {
        get
        {
            var periodMs = (CalculationPeriodEnd - CalculationPeriodStart).TotalMilliseconds;
            return (decimal)(periodMs / (1000 * 60 * 60));
        }
    }

    /// <summary>
    /// Determine if OEE requires attention based on thresholds
    /// </summary>
    /// <param name="thresholds">Optional custom thresholds</param>
    /// <returns>True if any metric is below threshold</returns>
    public bool RequiresAttention(OeeThresholds? thresholds = null)
    {
        var limits = thresholds ?? OeeThresholds.Default;

        return OeePercentage < limits.OeeThreshold ||
               AvailabilityPercentage < limits.AvailabilityThreshold ||
               PerformancePercentage < limits.PerformanceThreshold ||
               QualityPercentage < limits.QualityThreshold;
    }

    /// <summary>
    /// Identify which factor is the worst (constraining factor)
    /// </summary>
    /// <returns>The worst performing factor</returns>
    public OeeFactor GetWorstFactor()
    {
        var factors = new[]
        {
            (Factor: OeeFactor.Availability, Value: AvailabilityPercentage),
            (Factor: OeeFactor.Performance, Value: PerformancePercentage),
            (Factor: OeeFactor.Quality, Value: QualityPercentage)
        };

        return factors.OrderBy(f => f.Value).First().Factor;
    }

    /// <summary>
    /// Get improvement potential for each factor
    /// </summary>
    /// <param name="worldClassTargets">Optional world-class targets</param>
    /// <returns>Improvement potential breakdown</returns>
    public OeeImprovementPotential GetImprovementPotential(OeeTargets? worldClassTargets = null)
    {
        var targets = worldClassTargets ?? OeeTargets.WorldClass;

        return new OeeImprovementPotential(
            Math.Max(0, targets.AvailabilityTarget - AvailabilityPercentage),
            Math.Max(0, targets.PerformanceTarget - PerformancePercentage),
            Math.Max(0, targets.QualityTarget - QualityPercentage),
            Math.Max(0, (targets.AvailabilityTarget * targets.PerformanceTarget * targets.QualityTarget / 10000) - OeePercentage)
        );
    }

    /// <summary>
    /// Calculate what OEE would be if one factor improved
    /// </summary>
    /// <param name="factor">Factor to improve</param>
    /// <param name="newPercentage">New percentage for the factor</param>
    /// <returns>Simulated OEE percentage</returns>
    public decimal SimulateImprovement(OeeFactor factor, decimal newPercentage)
    {
        var (availability, performance, quality) = factor switch
        {
            OeeFactor.Availability => (newPercentage, PerformancePercentage, QualityPercentage),
            OeeFactor.Performance => (AvailabilityPercentage, newPercentage, QualityPercentage),
            OeeFactor.Quality => (AvailabilityPercentage, PerformancePercentage, newPercentage),
            _ => throw new ArgumentException($"Unknown OEE factor: {factor}", nameof(factor))
        };

        return (availability * performance * quality) / 10000;
    }

    /// <summary>
    /// Get OEE classification based on industry standards
    /// </summary>
    /// <returns>Classification description</returns>
    public string GetClassification()
    {
        var oee = OeePercentage;

        return oee switch
        {
            >= 85m => "World Class - Excellent performance",
            >= 65m => "Good - Acceptable performance",
            >= 40m => "Fair - Improvement needed",
            _ => "Poor - Significant improvement required"
        };
    }

    /// <summary>
    /// Get detailed breakdown of all components
    /// </summary>
    /// <returns>Complete OEE breakdown</returns>
    public OeeBreakdown GetBreakdown()
    {
        return new OeeBreakdown(
            OeePercentage,
            Availability.GetBreakdown(),
            Performance.GetBreakdown(),
            Quality.GetBreakdown(),
            GetWorstFactor().ToString(),
            GetClassification(),
            PeriodHours
        );
    }

    /// <summary>
    /// Create a summary object for reporting
    /// </summary>
    /// <returns>OEE summary for reports</returns>
    public OeeSummary ToSummary()
    {
        return new OeeSummary(
            Id,
            ResourceReference,
            CalculationPeriodStart,
            CalculationPeriodEnd,
            OeePercentage,
            AvailabilityPercentage,
            PerformancePercentage,
            QualityPercentage,
            GetWorstFactor().ToString(),
            RequiresAttention()
        );
    }

    /// <summary>
    /// Generate a unique OEE ID
    /// </summary>
    /// <param name="resourceReference">Resource identifier</param>
    /// <returns>Generated OEE ID</returns>
    private static string GenerateOeeId(string resourceReference)
    {
        var timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
        return $"OEE-{resourceReference}-{timestamp}";
    }

    /// <summary>
    /// String representation of the OEE calculation
    /// </summary>
    /// <returns>Formatted string representation</returns>
    public override string ToString()
    {
        return $"OEE: {OeePercentage:F1}% (A:{AvailabilityPercentage:F1}% × P:{PerformancePercentage:F1}% × Q:{QualityPercentage:F1}%)";
    }
}

/// <summary>
/// OEE factor enumeration
/// </summary>
public enum OeeFactor
{
    /// <summary>
    /// Availability factor
    /// </summary>
    Availability,

    /// <summary>
    /// Performance factor
    /// </summary>
    Performance,

    /// <summary>
    /// Quality factor
    /// </summary>
    Quality
}

/// <summary>
/// OEE thresholds for attention alerts
/// </summary>
/// <param name="OeeThreshold">Overall OEE threshold</param>
/// <param name="AvailabilityThreshold">Availability threshold</param>
/// <param name="PerformanceThreshold">Performance threshold</param>
/// <param name="QualityThreshold">Quality threshold</param>
public record OeeThresholds(
    decimal OeeThreshold,
    decimal AvailabilityThreshold,
    decimal PerformanceThreshold,
    decimal QualityThreshold
)
{
    /// <summary>
    /// Default thresholds for attention alerts
    /// </summary>
    public static readonly OeeThresholds Default = new(60m, 80m, 75m, 95m);
}

/// <summary>
/// World-class OEE targets
/// </summary>
/// <param name="AvailabilityTarget">Target availability percentage</param>
/// <param name="PerformanceTarget">Target performance percentage</param>
/// <param name="QualityTarget">Target quality percentage</param>
public record OeeTargets(
    decimal AvailabilityTarget,
    decimal PerformanceTarget,
    decimal QualityTarget
)
{
    /// <summary>
    /// World-class targets
    /// </summary>
    public static readonly OeeTargets WorldClass = new(90m, 95m, 99m);
}

/// <summary>
/// OEE improvement potential
/// </summary>
/// <param name="Availability">Availability improvement potential</param>
/// <param name="Performance">Performance improvement potential</param>
/// <param name="Quality">Quality improvement potential</param>
/// <param name="Overall">Overall OEE improvement potential</param>
public record OeeImprovementPotential(
    decimal Availability,
    decimal Performance,
    decimal Quality,
    decimal Overall
);

/// <summary>
/// Complete OEE breakdown
/// </summary>
/// <param name="OeePercentage">Overall OEE percentage</param>
/// <param name="Availability">Availability breakdown</param>
/// <param name="Performance">Performance breakdown</param>
/// <param name="Quality">Quality breakdown</param>
/// <param name="WorstFactor">Worst performing factor</param>
/// <param name="Classification">OEE classification</param>
/// <param name="PeriodHours">Period duration in hours</param>
public record OeeBreakdown(
    decimal OeePercentage,
    AvailabilityBreakdown Availability,
    PerformanceBreakdown Performance,
    QualityBreakdown Quality,
    string WorstFactor,
    string Classification,
    decimal PeriodHours
);

/// <summary>
/// OEE summary for reporting
/// </summary>
/// <param name="OeeId">OEE calculation ID</param>
/// <param name="ResourceReference">Resource identifier</param>
/// <param name="PeriodStart">Calculation period start</param>
/// <param name="PeriodEnd">Calculation period end</param>
/// <param name="OeePercentage">Overall OEE percentage</param>
/// <param name="AvailabilityPercentage">Availability percentage</param>
/// <param name="PerformancePercentage">Performance percentage</param>
/// <param name="QualityPercentage">Quality percentage</param>
/// <param name="WorstFactor">Worst performing factor</param>
/// <param name="RequiresAttention">Whether this calculation requires attention</param>
public record OeeSummary(
    string OeeId,
    string ResourceReference,
    DateTime PeriodStart,
    DateTime PeriodEnd,
    decimal OeePercentage,
    decimal AvailabilityPercentage,
    decimal PerformancePercentage,
    decimal QualityPercentage,
    string WorstFactor,
    bool RequiresAttention
);
